### PR TITLE
FEATURE: Make artist more creative

### PR DESCRIPTION
**Pull request by SamSaffron**
This allows for 2 big features:

1. Artist can ship up to 4 prompts for image generation
2. Artist can regenerate images cause it is aware of seed

This allows for iteration on images maintaining visual style

### CODE REVIEW

**Comment by SamSaffron** [comment]
work in progress, need to fix anthropic and tests...

**Comment by SamSaffron** [comment]
Note, this works properly on GPT-3.5 and 4, but we need a followup with Claude. The command framework is way to fragile and needs to be re-implemented for Claude

File: lib/modules/ai_bot/commands/image_command.rb [hunk 1]
```diff
@@ -53,55 +65,78 @@ def process(prompt:)
...
-                  DiscourseAi::Inference::StabilityGenerator.perform!(
-                    prompt,
-                    engine: engine,
-                    api_key: api_key,
-                    api_url: api_url,
-                  )
-              rescue => e
-                Rails.logger.warn("Failed to generate image for prompt #{prompt}: #{e}")
-              end
+      i = 0
```
Comment (martin-brennan) [inline lib/modules/ai_bot/commands/image_command.rb:68]: Can you just use `.map.with_index` below?

File: lib/modules/ai_bot/commands/image_command.rb [hunk 2]
```diff
@@ -53,55 +65,78 @@ def process(prompt:)
...
+      threads =
+        prompts.map do |prompt|
+          seed = seeds ? seeds[i] : nil
+          i += 1
+          Thread.new(seed, prompt) do |inner_seed, inner_prompt|
```
Comment (martin-brennan) [inline lib/modules/ai_bot/commands/image_command.rb:73]: I didn't now you could pass args to `Thread` like this, nice

File: lib/modules/ai_bot/commands/image_command.rb [hunk 3]
```diff
@@ -53,55 +65,78 @@ def process(prompt:)
...
-
-          show_progress(".", progress_caret: true) while !thread.join(2)
-
-          break if results
+          end
         end
+
+      while true
+        show_progress(".", progress_caret: true)
+        break if threads.all? { |t| t.join(2) }
```
Comment (martin-brennan) [inline lib/modules/ai_bot/commands/image_command.rb:95]: I also didn't know about the `limit` seconds for `join` :+1:

File: lib/modules/ai_bot/commands/image_command.rb [hunk 4]
```diff
@@ -53,55 +65,78 @@ def process(prompt:)
...
 
-      results[:artifacts].each_with_index do |image, i|
-        f = Tempfile.new("v1_txt2img_#{i}.png")
-        f.binmode
-        f.write(Base64.decode64(image[:base64]))
-        f.rewind
-        uploads << UploadCreator.new(f, "image.png").create_for(bot_user.id)
-        f.unlink
+      i = 0
+      results.each do |result|
```
Comment (martin-brennan) [inline lib/modules/ai_bot/commands/image_command.rb:107]: You can use `each_with_index` here too

File: lib/modules/ai_bot/commands/image_command.rb [hunk 5]
```diff
@@ -53,55 +65,78 @@ def process(prompt:)
...
+        result[:artifacts].each do |image|
+          f = Tempfile.new("v1_txt2img_#{i}.png")
```
Comment (martin-brennan) [inline lib/modules/ai_bot/commands/image_command.rb:109]: You can do:

```ruby
Tempfile.create("v1_txt2img_#{i}.png") do |file|
  file.binmode
  file.write(Base64.decode64(image[:base64]))
  file.rewind
  // upload stuff
end
```

Then it is automatically closed and unlinked at the end of the block. Ruby docs recommend doing the `close` and `unlink` in an `ensure` block if not using `create` https://devdocs.io/ruby~3.2/tempfile#class-Tempfile-label-Good+practices

**Review by martin-brennan — APPROVED** [review]
I don't have an API key for stability so can't try this locally, but the code looks good with some minor comments. Keen to try this on Meta :)

File: lib/modules/ai_bot/commands/image_command.rb (hunk 2 above)
Comment (SamSaffron) [reply lib/modules/ai_bot/commands/image_command.rb:73]: yeah, if you dont do that you can get some terrible concurrency bugs when stuff changes underneath you.

File: lib/modules/ai_bot/commands/image_command.rb (hunk 5 above)
Comment (SamSaffron) [reply lib/modules/ai_bot/commands/image_command.rb:109]: will change!

File: lib/modules/ai_bot/commands/image_command.rb (hunk 1 above)
Comment (SamSaffron) [reply lib/modules/ai_bot/commands/image_command.rb:68]: will change!

**Comment by SamSaffron** [comment]
thanks for the review :hugs:

**Pull request merged**